File: core/timezone.c

#include "graphserver.h"

Timezone*
tzNew( ) {
    Timezone* ret = (Timezone*)malloc(sizeof(Timezone));
    ret->head = NULL;
    
    return ret;
}

void
tzAddPeriod( Timezone* this, TimezonePeriod* period ) {
    if(!this->head) {
        this->head = period;
    } else {
        TimezonePeriod* prev = NULL;
        TimezonePeriod* curs = this->head;
        
        while(curs && period->begin_time > curs->end_time ) {
            prev = curs;
            curs = curs->next_period;
        }
        
        //link last and period; replace the head if necessary
        if(prev) {
            prev->next_period = period;
        } else {
            this->head = period;
        }
        
        //link period and curs
        period->next_period = curs;
        
    }
}

TimezonePeriod*
tzPeriodOf( Timezone* this, long time) {
  TimezonePeriod* period = this->head;

  while( period && period->end_time < time ) {
    period = period->next_period;
  }
  
  if( period && time < period->begin_time ) {
      return NULL;
  }
  
  return period;
}

int
tzUtcOffset( Timezone* this, long time) {
    //Returns seconds offset UTC for this timezone, at the given time
    
    TimezonePeriod* now = tzPeriodOf( this, time );
    
    if( !now ) {
        return -100*3600; //utc offset larger than any conceivable offset, as an error signal
    }
    
    return tzpUtcOffset( now );
}

int
tzTimeSinceMidnight( Timezone* this, long time ) {
    TimezonePeriod* now = tzPeriodOf( this, time );
    
    if( !now ) {
        return -1;
    }
    
    return (time+now->utc_offset)%SECS_IN_DAY;
}

TimezonePeriod*
tzHead( Timezone* this ) {
    return this->head;
}

void
tzDestroy( Timezone* this ) {
    TimezonePeriod* curs = this->head;
    TimezonePeriod* next;

    while(curs) {
      next = curs->next_period;
      tzpDestroy(curs);
      curs = next;
    }

    free(this);
}

TimezonePeriod*
tzpNew( long begin_time, long end_time, int utc_offset ) {
    TimezonePeriod* ret = (TimezonePeriod*)malloc(sizeof(TimezonePeriod));
    ret->begin_time    = begin_time;
    ret->end_time      = end_time;
    ret->utc_offset    = utc_offset;
    ret->next_period = NULL;

    return ret;
}

void
tzpDestroy( TimezonePeriod* this ) {
    free( this );
}

int
tzpUtcOffset( TimezonePeriod* this ) {
    return this->utc_offset;
}

int
tzpTimeSinceMidnight( TimezonePeriod* this, long time ) {
    return (time+this->utc_offset)%SECS_IN_DAY;
}

long
tzpBeginTime( TimezonePeriod* this ) {
    return this->begin_time;
}

long
tzpEndTime( TimezonePeriod* this ) {
    return this->end_time;
}

TimezonePeriod*
tzpNextPeriod(TimezonePeriod* this) {
    return this->next_period;
}

File: pygs/graphserver/core/timezone.py
from ..gsdll import c_long, c_void_p, cproperty, lgs, CShadow

class TimezonePeriod(CShadow):
    begin_time = cproperty(lgs.tzpBeginTime, c_long)
    end_time = cproperty(lgs.tzpEndTime, c_long)
    utc_offset = cproperty(lgs.tzpUtcOffset, c_long)

    def __init__(self, begin_time, end_time, utc_offset):
        self.soul = lgs.tzpNew(begin_time, end_time, utc_offset)

    @property
    def next_period(self):
        return TimezonePeriod.from_pointer(lgs.tzpNextPeriod(self.soul))

    def time_since_midnight(self, time):
        return lgs.tzpTimeSinceMidnight(self.soul, c_long(int(time)))

    def __getstate__(self):
        return (self.begin_time, self.end_time, self.utc_offset)

    def __setstate__(self, state):
        self.__init__(*state)


class Timezone(CShadow):
    head = cproperty(lgs.tzHead, c_void_p, TimezonePeriod)

    def __init__(self):
        self.soul = lgs.tzNew()

    def destroy(self):
        self.check_destroyed()

        self._cdel(self.soul)
        self.soul = None

    def add_period(self, timezone_period):
        lgs.tzAddPeriod(self.soul, timezone_period.soul)

    def period_of(self, time):
        tzpsoul = lgs.tzPeriodOf(self.soul, time)
        return TimezonePeriod.from_pointer(tzpsoul)

    def utc_offset(self, time):
        ret = lgs.tzUtcOffset(self.soul, time)

        if ret == -360000:
            raise IndexError("%d lands within no timezone period" % time)

        return ret

    def time_since_midnight(self, time):
        ret = lgs.tzTimeSinceMidnight(self.soul, c_long(int(time)))

        if ret == -1:
            raise IndexError("%d lands within no timezone period" % time)

        return ret

    @classmethod
    def generate(cls, timezone_string):
        ret = Timezone()

        timezone = pytz.timezone(timezone_string)
        tz_periods = zip(
            timezone._utc_transition_times[:-1], timezone._utc_transition_times[1:]
        )

        # exclude last transition_info entry, as it corresponds with the last utc_transition_time, and not the last period as defined by the last two entries
        for tz_period, (utcoffset, dstoffset, periodname) in zip(
            tz_periods, timezone._transition_info[:-1]
        ):
            period_begin, period_end = [
                calendar.timegm((x.year, x.month, x.day, x.hour, x.minute, x.second))
                for x in tz_period
            ]
            period_end -= 1  # period_end is the last second the period is active, not the first second it isn't
            utcoffset = utcoffset.days * 24 * 3600 + utcoffset.seconds

            ret.add_period(TimezonePeriod(period_begin, period_end, utcoffset))

        return ret

    def __getstate__(self):
        ret = []
        curs = self.head
        while curs:
            ret.append(curs.__getstate__())
            curs = curs.next_period
        return ret

    def __setstate__(self, state):
        self.__init__()
        for tzpargs in state:
            self.add_period(TimezonePeriod(*tzpargs))

    def expound(self):
        return "Timezone"


Timezone._cdel = lgs.tzDestroy
